File: sol/951/951.py
class Solution:
    def flipEquiv(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        if root1 is None or root2 is None:
            return root1 is None and root2 is None
        if root1.val != root2.val:
            return False
        l1, r1 = root1.left, root1.right
        if r1 is None or (l1 is not None and l1.val > r1.val):
            l1, r1 = r1, l1
        l2, r2 = root2.left, root2.right
        if r2 is None or (l2 is not None and l2.val > r2.val):
            l2, r2 = r2, l2
        return self.flipEquiv(l1, l2) and self.flipEquiv(r1, r2)

File: sol/951/951.cpp
struct Solution {
    bool flipEquiv(TreeNode* root1, TreeNode* root2) {
        if (root1 == nullptr || root2 == nullptr){
            return root1 == nullptr && root2 == nullptr;
        }
        if (root1->val != root2->val){
            return false;
        }
        TreeNode *l1 = root1->left, *r1 = root1->right, *l2 = root2->left, *r2 = root2->right;
        if (r1 == nullptr || (l1 != nullptr && l1->val > r1->val)){
            swap(l1, r1);
        }
        if (r2 == nullptr || (l2 != nullptr && l2->val > r2->val)){
            swap(l2, r2);
        }
        return this->flipEquiv(l1, l2) && this->flipEquiv(r1, r2);
    }
};
